### PR TITLE
Downsampling supports date_histogram with tz

### LINKED ISSUE #101309: Downsample time zone usability issue
The fact that downsampling rolls data up using UTC as fixed timezone is a known limitation: 
https://www.elastic.co/guide/en/elasticsearch/reference/8.10/downsampling.html#querying-downsampled-indices-notes

However when querying a tsdb data stream this can still come as a surprise. As long as downsampling hasn't been execute or configured or querying only the non downsampled part of a data stream (recent data isn't usually downsampled), the `date_histogram` aggregation works fine with any configured time zone. However when a search with a `date_histogram` does hit a shard that contains downsampled data, then the search fails with the following error: `Field [@timestamp] of type [date] is not supported for aggregation [timeseries] with timezone [America/Los_Angeles]`. This can be confusing as the user may not know that the data stream that is being queried has been configured to downsample at some point in time. And this failure is unlikely to happens when recent data is queried. Kibana adds timezone parameter automatically to the `date_histogram` aggregation, so often this outside of the control of the user (especially with builtin dashboards).

There are various options we can explore in order to mitigate this usability issue. This issue serves as a placeholder. Note that this isn't a `date_histogram` aggregation issue only. This is an issue that will need to be addressed in es-ql too.

### PR DESCRIPTION
**Pull request by kkrik-es**
This comes with caveats, for downsampled indexes at intervals more than 15 minutes. For instance,
 - 1-hour downsampling may report time buckets for 1-hour histograms on XX:30 or XX:15, instead of XX:00 (HH:MM) 
 - 1-day downsampling may report time buckets for daily histograms on XX:XX instead of 00:00 (HH:MM)

For more details, see Option 4 in [Downsampling and timezones](https://docs.google.com/document/d/1NQfdwsxCtW2dGPKBZD_TF8bN1Rq2XUvBadYhmjntdos/) doc.

Fixes #101309

### CODE REVIEW

**Comment by kkrik-es** [comment]
@elasticsearchmachine run elasticsearch-ci/part-4

**Comment by kkrik-es** [comment]
@martijnvg this has been updated to offset buckets and tag responses when this happens. Ptal.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateHistogramAggregationBuilder.java [hunk 1]
```diff
@@ -406,23 +408,42 @@ protected ValuesSourceAggregatorFactory innerBuild(
...
+            if (tz != null && ZoneId.of("UTC").equals(tz) == false && field().equals(DataStreamTimestampFieldMapper.DEFAULT_PATH)) {
+
+                // Get the downsampling interval.
+                DateHistogramInterval interval = new DateHistogramInterval(downsamplingInterval);
+                long downsamplingResolution = interval.estimateMillis();
+                long aggregationResolution = dateHistogramInterval.getAsFixedInterval().estimateMillis();
+
+                // If the aggregation resolution is not a multiple of the downsampling resolution, the reported time for each
+                // bucket needs to be shifted by the mod - in addition to rounding that's applied as usual.
+                long aggregationOffset = SimpleTimeZone.getTimeZone(tz).getOffset(aggregationResolution) % downsamplingResolution;
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateHistogramAggregationBuilder.java:441]: can the comments here also include that using intervals in ms (from `aggregationResolution`) as utc timestamps okay?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateHistogramAggregationBuilder.java [hunk 2]
```diff
@@ -406,23 +408,42 @@ protected ValuesSourceAggregatorFactory innerBuild(
...
+                        + " is not supported for aggregation ["
+                        + getName()
+                        + "] with interval type ["
+                        + dateHistogramIntervalType.getPreferredName()
+                        + "]"
+                );
+            }
+
+            // Downsampled data in time-series indexes contain aggregated values that get calculated over UTC-based intervals.
+            // When they get aggregated using a different timezone, the resulting buckets may need to be offset to account for
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateHistogramAggregationBuilder.java:428]: Maybe also describe that this method is best effort. And for example if time zones haveDST shifts then these aren't visible.

File: x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/TransportDownsampleAction.java [hunk 3]
```diff
@@ -888,6 +890,7 @@ public ClusterState execute(ClusterState currentState) {
                             Settings.builder()
                                 .put(downsampleIndex.getSettings())
                                 .put(IndexMetadata.INDEX_DOWNSAMPLE_STATUS.getKey(), DownsampleTaskStatus.SUCCESS)
+                                .put(IndexMetadata.INDEX_DOWNSAMPLE_INTERVAL.getKey(), downsamplingInterval)
```
Comment (martijnvg) [inline x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/TransportDownsampleAction.java:893]: This on its own is a good change.

File: docs/reference/data-streams/downsampling.asciidoc [hunk 4]
```diff
@@ -135,7 +135,28 @@ downsampled.
...
+the results include the field `downsampled_results_offset: true`, to indicate that
+the time buckets are shifted. This can be avoided if a downsampling interval of 15
+minutes is used, as it allows properly calculating hourly values for the shifted
+buckets.
+
+** Date histograms at intervals that are multiples of a day are similarly
+affected, in case downsampling aggregates values per day. In this case, the
+beginning of each day is always calculated at UTC when generated the downsampled
+values, so the time buckets need to be shifted, e.g. reported as
+`2020-01-01T19:00:00.000` instead of `2020-03-07T00:00:00.000`. The field
```
Comment (martijnvg) [inline docs/reference/data-streams/downsampling.asciidoc:154]: Should `2020-01-01T19:00:00.000` be `2020-03-07T19:00:00.000` instead? Maybe also mention the spefic time zone here? (east coast?)

File: docs/reference/data-streams/downsampling.asciidoc [hunk 5]
```diff
@@ -135,7 +135,28 @@ downsampled.
...
+`downsampled_results_offset: true` is added in this case too.
+
+** Daylight savings and similar peculiarities around timezones affect
```
Comment (martijnvg) [inline docs/reference/data-streams/downsampling.asciidoc:157]: I think we should report here the daylight savings can be ignored ignored / lost? (e.g downsample interval is 1 day?)

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java [hunk 6]
```diff
@@ -1277,6 +1272,13 @@ public String toString() {
         Property.InternalIndex
     );
 
+    public static final Setting<String> INDEX_DOWNSAMPLE_INTERVAL = Setting.simpleString(
+        INDEX_DOWNSAMPLE_INTERVAL_KEY,
+        "",
+        Property.IndexScope,
+        Property.InternalIndex
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:1279]: I think this should be a private setting as well? Only the downsample api should be allowed to set it.

File: x-pack/plugin/downsample/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/downsample/30_date_histogram.yml [hunk 7]
```diff
@@ -247,3 +229,477 @@ setup:
...
+                field: "@timestamp"
+                fixed_interval: 1h
+                time_zone: "America/New_York"
+          query:
+            range:
+              "@timestamp":
+                gt: "2021-04-28T15:00:00Z"
+                lt: "2021-04-29T15:00:00Z"
+
+  - match: { hits.total.value: 4 }
```
Comment (martijnvg) [inline x-pack/plugin/downsample/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/downsample/30_date_histogram.yml:311]: I think we should add an assertion here and that checks that `downsampled_results_offset` key is unset / not defined.  And also for other searches where we don't expect that downsampled_results_offset to be returned.

**Review by martijnvg — COMMENTED** [review]
I left a couple of comments, but looks good.

**Review by martijnvg — APPROVED** [review]
LGTM

File: x-pack/plugin/downsample/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/downsample/30_date_histogram.yml (hunk 7 above)
Comment (kkrik-es) [reply x-pack/plugin/downsample/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/downsample/30_date_histogram.yml:311]: Done.

**Pull request merged**